java/arcs/jvm/host/JvmSchedulerProvider.kt

```diff
@@ -0,0 +1,86 @@
+/*
+ * Copyright 2020 Google LLC.
+ *
+ * This code may only be used under the BSD style license found at
+ * http://polymer.github.io/LICENSE.txt
+ *
+ * Code distributed by Google as part of this project is also subject to an additional IP rights
+ * grant found at
+ * http://polymer.github.io/PATENTS.txt
+ */
+
+package arcs.jvm.host
+
+import arcs.core.host.SchedulerProvider
+import arcs.core.util.Scheduler
+import arcs.jvm.util.JvmTime
+import java.util.concurrent.Executors
+import kotlin.coroutines.CoroutineContext
+import kotlinx.coroutines.CoroutineDispatcher
+import kotlinx.coroutines.CoroutineName
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.asCoroutineDispatcher
+
+/**
+ * Implementation of a [SchedulerProvider] for the Java Virtual Machine (including Android).
+ *
+ * A new [Scheduler] is returned for each unique Arc ID encountered, and each [Scheduler] will be
+ * associated with a single-threaded [CoroutineDispatcher]. However, those dispatchers will be
+ * allocated to schedulers from a finite pool of dispatchers in a round-robin fashion - the size
+ * of that pool is defined by [maxThreadCount].
+ *
+ * Each [ArcHost] should generally use a separate [JvmSchedulerProvider].
+ */
+class JvmSchedulerProvider(
+    private val baseCoroutineContext: CoroutineContext,
+    private val maxThreadCount: Int =
+        maxOf(1, Runtime.getRuntime().availableProcessors() / 2),
+    private val threadPriority: Int = DEFAULT_THREAD_PRIORITY
+) : SchedulerProvider {
+    private val dispatchers = mutableListOf<CoroutineDispatcher>()
+    private val schedulersByArcId = mutableMapOf<String, Scheduler>()
+
+    @Synchronized
+    override fun invoke(arcId: String): Scheduler {
+        schedulersByArcId[arcId]?.let { return it }
+
+        val dispatcher = if (dispatchers.size == maxThreadCount) {
+            dispatchers[schedulersByArcId.size % maxThreadCount]
+        } else {
+            Executors
+                .newSingleThreadExecutor {
+                    // TODO(jasonwyatt): Create and cache this thread outside of the thread factory,
+                    //  so we can ensure we always return it - rather than creating new threads
+                    //  when/if the on the executor was created with failed or was killed.
+                    Thread(it).apply {
+                        priority = threadPriority
+                    }
+                }
+                .asCoroutineDispatcher()
+                .also { dispatchers.add(it) }
+        }
+
+        val schedulerParentJob = Job(baseCoroutineContext[Job])
+        schedulerParentJob.invokeOnCompletion {
+            synchronized(this@JvmSchedulerProvider) {
+                schedulersByArcId.remove(arcId)
+            }
+        }
+
+        val schedulerContext = baseCoroutineContext +
+            schedulerParentJob +
```

Comment:
I'm not sure I follow, I'm pretty sure we _want_ the parent to get cancelled/crashed when an exception happens in one of the child jobs in this situation.

Reply:
Do to mean, if a single say, onUpdate() callback throws an exception, then other particles in the same Arc that have pending jobs will be cancelled?  This sounds like a violation of the kinds of isolation Arcs is supposed to provide. I don't think ParticleA should be able invalidate work that ParticleB has already successfully scheduled.

Reply:
"Pending jobs" ?    particles themselves don't enqueue any actual work into the scheduler.  Their callbacks _are_ the work.    The only other work enqueued into the scheduler would be incoming model updates/operation updates from the Store.     So in this situation, I think crashing within one of the jobs SHOULD crash the process.

Reply:
I was thinking along the lines of handle.store() operations. Is applyOp() not going also be run on a scheduler?

Reply:
no, because it's already being effectively run on the scheduler - since all executions of `applyOp` need to occur within one of the callbacks (or on the scheduler itself) to begin with.